rholang/src/test/scala/coop/rchain/rholang/interpreter/accounting/CostAccountingPropertyTest.scala

```diff
@@ -1,27 +1,115 @@
 package coop.rchain.rholang.interpreter.accounting
+import java.nio.file.Files
+
+import cats.Id
+import cats.effect.Sync
+import coop.rchain.crypto.hash.Blake2b512Random
+import cats.implicits._
+import coop.rchain.models._
+import coop.rchain.rholang.interpreter.Runtime.{RhoContext, RhoISpace}
+import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 import coop.rchain.rholang.{PrettyPrinted, ProcGen}
-import coop.rchain.rholang.interpreter.Interpreter
-import coop.rchain.rholang.syntax.rholang_mercury.Absyn.Proc
+import coop.rchain.rholang.interpreter._
+import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{PPar, Proc}
 import coop.rchain.rholang.syntax.rholang_mercury.PrettyPrinter
+import coop.rchain.rspace.history.Branch
+import coop.rchain.rspace.{Context, RSpace}
+import monix.eval.Task
 import org.scalacheck.Arbitrary
 import org.scalacheck.Test.Parameters
 import org.scalatest.{FlatSpec, Matchers}
 import org.scalatest.prop.PropertyChecks
 
+import scala.concurrent.duration._
+import monix.execution.Scheduler.Implicits.global
+
 class CostAccountingPropertyTest extends FlatSpec with PropertyChecks with Matchers {
+  import CostAccountingPropertyTest._
+
   implicit val params: Parameters = Parameters.defaultVerbose.withMinSuccessfulTests(1000)
   implicit val procArbitrary = Arbitrary(
     ProcGen.topLevelGen(5).map(PrettyPrinted[Proc](_, PrettyPrinter.print))
   )
 
+  implicit val taskExecutionDuration = 5.seconds
+
   def cost(proc: Proc): Cost = Cost(Interpreter.buildPar(proc).apply)
 
   behavior of "Cost accounting in Reducer"
 
-  it should "cost of substitution is never negative" in {
+  it should "have positive cost of substitution" in {
     forAll { p: PrettyPrinted[Proc] =>
       cost(p.value).value should be >= 0L
```

Comment:
I think we should have some test saying that overall cost is indeed `> 0` for every PProc possible (even the Nil one shouldn't be for free)

We should also think of other properties that we could test about the cost - some more advanced and closer to our particular domain.

Reply:
I agree that we need more tests. I'll talk to Mateusz when he is back. Also if you or anybody has ideas about tests that might be relevant here, please shoot.

Reply:
There is not much more properties that we can test. From the high level view we have `p` and `q` processes and respective `cost(p)` and `cost(q)` . You can  test:

- `cost(p | q) == cost(p) + cost(q)`
- `cost(p) + cost(q) == cost(q) + cost(p)`
- `cost(p | Nil) == cost(p)`

I can't think of any more tests.

Reply:
I think one more property that should hold is:

`cost(p(q1, q2, .., qn)) > cost(q1) + ... + cost(qn)`

where `p(q1, q2, .., qn)` means `q...` are all the children of `p`.

I realize though testing this might be a separete jira.

@goral09 please let us know if you think the property makes sense :)

Reply:
And what this `parent`-`child` relation implies? I don't understand it.

Reply:
E.g. in `EPlus(GInt(1), GInt(2))` the GInt-s are children of EPlus. Or in 
`for (x <- z)  P` the children are `z` and `P`.

Reply:
It would test that cost of `p` is non-zero. Which is not true for in all the cases and in those which it is we already test it elsewhere.

Reply:
Allowing zero cost for any `p` is dangerous, becuase then I can stuff a lot of those into a container that doesn't add its own cost, like [p, p, p, p]... and troll the chain.

Reply:
`we already test it elsewhere` - not with generative tests though?

Reply:
> Allowing zero cost for any `p` is dangerous, becuase then I can stuff a lot of those into a container that doesn't add its own cost, like [p, p, p, p]... and troll the chain.

The cost of calling contracts defined in `SystemProcesses` is zero: `std[Out|Err](Ack)` and crypto methods. Although I think we should discuss again the rationale behind allowing for crypto methods be free. I will raise that issue.

Reply:
> `we already test it elsewhere` - not with generative tests though?

No, but since some of the methods are free those tests will fail.